pkg/kubectl/cmd/util/shortcut_restmapper.go

```diff
@@ -136,3 +159,109 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
 
 	return resource
 }
+
+func (e *shortcutExpander) wrapError(err error, resource schema.GroupVersionResource) error {
+	switch er := err.(type) {
+	case *meta.NoResourceMatchError:
+		return &NoResourceWithCommendError{
+			NoResourceMatchError: er,
+			Recommends:           e.similarResources(resource),
+		}
+	}
+
+	return err
+}
+
+func (e *shortcutExpander) similarResources(resource schema.GroupVersionResource) []string {
+	if len(e.cachedShortcuts) == 0 {
+		return []string{}
+	}
+
+	allNames := map[string]bool{}
+	for _, shortcut := range e.cachedShortcuts {
+		allNames[shortcut.LongForm.String()] = true
+		allNames[shortcut.ShortForm.String()] = true
+		allNames[shortcut.LongForm.Resource] = true
+		allNames[shortcut.ShortForm.Resource] = true
+	}
+
+	results := map[int][]string{}
+	gr := resource.GroupResource()
+	res := (&gr).String()
+	max := len(res) / 2
+
+	for name := range allNames {
+		// TODO: paraellel it
+		dist := editDistance(res, name)
+		if dist <= max {
+			results[dist] = append(results[dist], name)
+		}
+	}
+
+	for i := 1; i <= max; i++ {
+		if result, ok := results[i]; ok {
+			return result
+		}
+	}
+
+	return []string{}
+}
+
+func editDistance(s1, s2 string) int {
```

Comment:
Don’t we already have a dependency on an edit distance library via spf13/cobra, which already suggests this for command names?  If so, use that library.

Reply:
@smarterclayton yes, kubectl is using cobra. to my best knowledge, cobra deals with predefined subcommands, and suggests names for wrong subcommands. this patch also suggests wrong resource names, which are dynamic. They have a levenshteinDistance method, but it's not intended for consumption, also it's private.

Reply:
Can you also move it to a separate file (`helpers.go`) , please?